### PR TITLE
sof: spinlock: Fix several compile errors related to spinlock debug

### LINKED ISSUE #2015: [BUG] FW Compile error when enabling verbose spinlock debug
**Describe the bug**

When selecting DEBUG and DEBUG_LOCKS_VERBOSE on the i.MX platform I get the following compile error:

```
[trimmed]/src/include/sof/spinlock.h:87:9: error: 'esai[i].lock' is a pointer; did you mean to use '->'?
   (lock)->user = __LINE__; \
```

(two times in my case, within the macro `spinlock_init`; I think dependent on platform this macro may be called a different number of times but I believe the error is actually not platform-specific)

**To Reproduce**
Steps to reproduce the behavior: (e.g. list commands or actions used to reproduce the bug)
* Generate default configuration: `make imx8_defconfig` (replace with your platform as I don't think this is platform specific)
* Select, in addition to the defconfig, DEBUG, DEBUG_LOCKS ~and DEBUG_LOCKS_VERBOSE~ (also occurs without selecting DEBUG_LOCKS_VERBOSE).
* Attempt build: `make -j8`

**Reproduction Rate**
100% (build error)

**Expected behavior**
Firmware should build successfully

**Impact**
What impact does this issue have on your progress (e.g., annoyance, showstopper)
Annoyance

**Environment**
1) Branch name and commit hash of the 2 repositories: sof (firmware/topology) and linux (kernel driver).
    * Kernel: (doesn't matter)
    * SOF: 78b022bf (master)
2) Name of the platform(s) on which the bug is observed.
    * Platform: imx8 (though I believe this is not platform-specific)

### PR DESCRIPTION
**Pull request by paulstelian97**
Signed-off-by: Paul Olaru <paul.olaru@nxp.com>

Fixes #2015

### CODE REVIEW

File: src/include/sof/spinlock.h [hunk 1]
```diff
@@ -84,7 +85,7 @@ extern uint32_t lock_dbg_user[DBG_LOCK_USERS];
 #define spinlock_init(lock) \
 	do { \
 		arch_spinlock_init(lock); \
-		(lock)->user = __LINE__; \
+		(*lock)->user = __LINE__; \
```
Comment (jajanusz) [inline src/include/sof/spinlock.h]: Lovely macros again. If I see correctly `lock` is pointer or pointer to pointer, so we can easily turn this macro into static inline and get type-checking.

Comment (paulstelian97) [reply src/include/sof/spinlock.h]: Should I do it in this PR (separate commit) or leave it as an issue so it can be discussed more?

**Comment by paulstelian97** [comment]
@lgirdwood Should I paste the messages coming from my platform specific uses of the macro `spinlock_init`? Or just the ones in shared code (ll_schedule_domain.h)?

**Comment by paulstelian97** [comment]
Pasting here first, commit afterwards:
```
/work/repos/sof/src/platform/imx8/lib/dai.c: In function 'dai_init':
/work/repos/sof/src/include/sof/spinlock.h:87:9: error: 'esai[i].lock' is a pointer; did you mean to use '->'?
   (lock)->user = __LINE__; \
         ^~
/work/repos/sof/src/platform/imx8/lib/dai.c:84:3: note: in expansion of macro 'spinlock_init'
   spinlock_init(&esai[i].lock);
   ^~~~~~~~~~~~~
/work/repos/sof/src/include/sof/spinlock.h:87:9: error: 'sai[i].lock' is a pointer; did you mean to use '->'?
   (lock)->user = __LINE__; \
         ^~
/work/repos/sof/src/platform/imx8/lib/dai.c:87:3: note: in expansion of macro 'spinlock_init'
   spinlock_init(&sai[i].lock);
   ^~~~~~~~~~~~~
[...]
/work/repos/sof/src/include/sof/schedule/ll_schedule_domain.h: In function 'domain_init':
/work/repos/sof/src/include/sof/spinlock.h:87:9: error: 'domain->lock' is a pointer; did you mean to use '->'?
   (lock)->user = __LINE__; \
         ^~
/work/repos/sof/src/include/sof/schedule/ll_schedule_domain.h:71:2: note: in expansion of macro 'spinlock_init'
  spinlock_init(&domain->lock);
  ^~~~~~~~~~~~~
[...]
/work/repos/sof/src/platform/imx8/lib/dma.c: In function 'edma_init':
/work/repos/sof/src/include/sof/spinlock.h:87:9: error: 'dma[i].lock' is a pointer; did you mean to use '->'?
   (lock)->user = __LINE__; \
         ^~
/work/repos/sof/src/platform/imx8/lib/dma.c:54:3: note: in expansion of macro 'spinlock_init'
   spinlock_init(&dma[i].lock);
   ^~~~~~~~~~~~~
```

Should I put these in the commit message? @lgirdwood

**Comment by paulstelian97** [comment]
@lgirdwood I believe I managed to make the commit message a little too verbose, at least given my taste. But if including the error messages is helpful, then sure, did that.

**Comment by dbaluta** [comment]
@paul commit is Ok. I would suggest to create two commits each fixing a thing.

Also, macro is fixed now we can leave the transformation to an inline function for later. Just create an issue for it.

**Comment by paulstelian97** [comment]
> @paul commit is Ok. I would suggest to create two commits each fixing a thing.
> 
> Also, macro is fixed now we can leave the transformation to an inline function for later. Just create an issue for it.

Will split the commits, I worried about the usual rule of not splitting compile error fixes due to Git Bisect. Guess this rule only applies to the default and maybe to other usual configs.

File: src/include/sof/spinlock.h [hunk 2]
```diff
@@ -15,6 +15,7 @@
 
 #include <arch/spinlock.h>
 #include <sof/drivers/interrupt.h>
+#include <sof/trace/trace.h>
 #include <config.h>
```
Comment (dbaluta) [inline src/include/sof/spinlock.h]: Why do we need this include added in this patch?

Comment (dbaluta) [reply src/include/sof/spinlock.h]: no its fine

**Comment by dbaluta** [comment]
@jajanusz @tlauda @lgirdwood we need your blessing on this.

File: src/include/sof/spinlock.h (hunk 1 above)
Comment (tlauda) [reply src/include/sof/spinlock.h]: Approving this one, but let's follow up with what @jajanusz suggested.

Comment (jajanusz) [reply src/include/sof/spinlock.h]: Feel free to do it in separeate commit/PR

**Pull request merged**